File: OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/model/OSOutcomeSource.java
package com.onesignal.outcomes.model;

import android.support.annotation.Nullable;

import org.json.JSONException;
import org.json.JSONObject;

public class OSOutcomeSource {

    private static final String DIRECT = "direct";
    private static final String INDIRECT = "indirect";

    @Nullable
    private OSOutcomeSourceBody directBody;
    @Nullable
    private OSOutcomeSourceBody indirectBody;

    public OSOutcomeSource(@Nullable OSOutcomeSourceBody directBody, @Nullable OSOutcomeSourceBody indirectBody) {
        this.directBody = directBody;
        this.indirectBody = indirectBody;
    }

    public JSONObject toJSONObject() throws JSONException {
        JSONObject json = new JSONObject();
        if (directBody != null)
            json.put(DIRECT, directBody.toJSONObject());
        if (indirectBody != null)
            json.put(INDIRECT, indirectBody.toJSONObject());
        return json;
    }

    public OSOutcomeSourceBody getDirectBody() {
        return directBody;
    }

    public OSOutcomeSource setDirectBody(@Nullable OSOutcomeSourceBody directBody) {
        this.directBody = directBody;
        return this;
    }

    public OSOutcomeSourceBody getIndirectBody() {
        return indirectBody;
    }

    public OSOutcomeSource setIndirectBody(@Nullable OSOutcomeSourceBody indirectBody) {
        this.indirectBody = indirectBody;
        return this;
    }

    @Override
    public String toString() {
        return "OSOutcomeSource{" +
                "directBody=" + directBody +
                ", indirectBody=" + indirectBody +
                '}';
    }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/outcomes/domain/OSOutcomeSource.kt
package com.onesignal.outcomes.domain

import com.onesignal.outcomes.OSOutcomeConstants
import org.json.JSONException
import org.json.JSONObject

class OSOutcomeSource(var directBody: OSOutcomeSourceBody?,
                      var indirectBody: OSOutcomeSourceBody?) {
    @Throws(JSONException::class)
    fun toJSONObject(): JSONObject {
        val json = JSONObject()
        directBody?.let {
            json.put(OSOutcomeConstants.DIRECT, it.toJSONObject())
        }
        indirectBody?.let {
            json.put(OSOutcomeConstants.INDIRECT, it.toJSONObject())
        }
        return json
    }

    fun setDirectBody(directBody: OSOutcomeSourceBody?) = this.apply {
        this.directBody = directBody
    }

    fun setIndirectBody(indirectBody: OSOutcomeSourceBody?) = this.apply {
        this.indirectBody = indirectBody
    }

    override fun toString(): String {
        return "OSOutcomeSource{" +
                "directBody=" + directBody +
                ", indirectBody=" + indirectBody +
                '}'
    }
}